Spigot-Server-Patches/0248-Allow-biome-replacement-in-config-file.patch

```diff
@@ -0,0 +1,234 @@
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/Biomes.java b/src/main/java/net/minecraft/server/Biomes.java
+index 4165febc5..1faa8ae45 100644
+--- a/src/main/java/net/minecraft/server/Biomes.java
++++ b/src/main/java/net/minecraft/server/Biomes.java
+@@ -1,72 +1,85 @@
+ package net.minecraft.server;
+ 
++// Paper start
++import com.destroystokyo.paper.PaperConfig;
++// Paper end
++
+ public abstract class Biomes {
+ 
+-    public static final BiomeBase a;
+-    public static final BiomeBase b;
+-    public static final BiomeBase c;
+-    public static final BiomeBase d;
+-    public static final BiomeBase e;
+-    public static final BiomeBase f;
+-    public static final BiomeBase g;
+-    public static final BiomeBase h;
+-    public static final BiomeBase i;
+-    public static final BiomeBase j;
+-    public static final BiomeBase k;
+-    public static final BiomeBase l;
+-    public static final BiomeBase m;
+-    public static final BiomeBase n;
+-    public static final BiomeBase o;
+-    public static final BiomeBase p;
+-    public static final BiomeBase q;
+-    public static final BiomeBase r;
+-    public static final BiomeBase s;
+-    public static final BiomeBase t;
+-    public static final BiomeBase u;
+-    public static final BiomeBase v;
+-    public static final BiomeBase w;
+-    public static final BiomeBase x;
+-    public static final BiomeBase y;
+-    public static final BiomeBase z;
+-    public static final BiomeBase A;
+-    public static final BiomeBase B;
+-    public static final BiomeBase C;
+-    public static final BiomeBase D;
+-    public static final BiomeBase E;
+-    public static final BiomeBase F;
+-    public static final BiomeBase G;
+-    public static final BiomeBase H;
+-    public static final BiomeBase I;
+-    public static final BiomeBase J;
+-    public static final BiomeBase K;
+-    public static final BiomeBase L;
+-    public static final BiomeBase M;
+-    public static final BiomeBase N;
+-    public static final BiomeBase O;
+-    public static final BiomeBase P;
+-    public static final BiomeBase Q;
+-    public static final BiomeBase R;
+-    public static final BiomeBase S;
+-    public static final BiomeBase T;
+-    public static final BiomeBase U;
+-    public static final BiomeBase V;
+-    public static final BiomeBase W;
+-    public static final BiomeBase X;
+-    public static final BiomeBase Y;
+-    public static final BiomeBase Z;
+-    public static final BiomeBase aa;
+-    public static final BiomeBase ab;
+-    public static final BiomeBase ac;
+-    public static final BiomeBase ad;
+-    public static final BiomeBase ae;
+-    public static final BiomeBase af;
+-    public static final BiomeBase ag;
+-    public static final BiomeBase ah;
+-    public static final BiomeBase ai;
+-    public static final BiomeBase aj;
+-    public static final BiomeBase ak;
++    // Paper start - Remove final on biome members
```

Comment:
Why was this done? It doesn't look needed. I don't see any assignment. Please remove all of this

Reply:
I was hoping for a plug-in API later on that would make it easy to modify this which if its final would prevent a consistent plug-in API from being formed and any plug-ins would have to resort to reflection on minecraft specific obfuscation code. But I'd be happy to remove this if its needed to.

Reply:
Also the entire code won't work work if its final. Biomes.java is initialized at the very start of the entire program right when "main" is called. Therefore the config has not remotely been loaded yet. The way I work around this is by loading vanilla Biome values into the fields and after the config is loaded re-load the values into the fields properly in-case the config overrides anything.

That's why I moved the code to "setupBiomes()" because its a method that's called once at the very start of the entire program before any lines of code are executed, and once after the config file is loaded and parsed.

If the members are made final, then the method would not be able to be called anymore than once and the whole feature wouldn't work.

Reply:
Unless Java is doing something extremely weird in execution order.... then your code is still behaving the same as if you didnt make this change...

It was all in a static initialization block.

You moved it to all be inside of a method, but then that method is immediately executed inside of a static initialization block.....
I don't think the behavior is any different than if you removed all of this diff.

@kashike @electronicboy  @zachbr thoughts? Does wrapping the code in a method change init behavior here?

Reply:
I think that there is a change in behavior, but that's only because the setup biomes stuff was called in multiple places here, these fields are effectively initialized twice due to this. only real fix for this would be to ensure that the paper config is loaded up earlier, but that would also need us to move stuff like spigots config initialization too, which I was looking into but they use the bukkit logger; so we'd also need to patch their config for that too